powersimdata/design/generation/curtailment.py

```diff
@@ -0,0 +1,124 @@
+import pandas as pd
+
+from powersimdata.scenario.scenario import Scenario
+
+default_pmin_dict = {
+    "coal": None,
+    "dfo": 0,
+    "geothermal": 0.95,
+    "hydro": None,
+    "ng": 0,
+    "nuclear": 0.95,
+    "other": 0,
+    "solar": 0,
+    "wind": 0,
+    "wind_offshore": 0,
+}
+profile_methods = {
+    "hydro": "get_hydro",
+    "solar": "get_solar",
+    "wind": "get_wind",
+    "wind_offshore": "get_wind",
+}
+
+
+def temporal_curtailment(
+    scenario, pmin_by_type=None, pmin_by_id=None, curtailable={"solar", "wind"}
+):
+    """Calculate the minimum share of potential renewable energy that will be curtailed
+        due to supply/demand mismatch, assuming no storage is present.
+
+    :param powersimdata.scenario.scenario.Scenario scenario: scenario instance.
+    :param dict/pandas.Series pmin_by_type: Mapping of types to Pmin assumptions. Values
+        between 0 and 1 (inclusive) are treated as shares of Pmax, and None values
+        either maintain given Pmin values (for dispatchable resources) or track profiles
+        (for profile resources, e.g. hydro).
+    :param dict/pandas.Series pmin_by_id: Mapping of IDs to Pmin assumptions, as an
+        override to the default behavior for that plant type. Values between 0 and 1
+        (inclusive) are treated as shares of Pmax, and None values either maintain given
+        Pmin values (for dispatchable resources) or track profiles
+        (for profile resources, e.g. hydro).
+    :param iterable curtailable: resource types which can be curtailed.
+    :return: (*float*) -- share of curtailable resources that will be curtailed.
+    :raises TypeError: if inputs do not mach specified type.
+    :raises ValueError: if any entries in curtailable or keys in pmin_by_type are not
+        types in the grid, any keys in pmin_by_id are not plant IDs in the Grid,
+        or any values in pmin_by_type/pmin_by_id are not in the range [0, 1] or None.
+    """
+    if not isinstance(scenario, Scenario):
+        raise TypeError("scenario must be a Scenario")
+    if pmin_by_type is None:
+        pmin_by_type = {}
+    if pmin_by_id is None:
+        pmin_by_id = {}
```

Comment:
Should we set the default value to an empty `dict` instead of `None` since it is overwritten above

Reply:
I think @jon-hagg suggested we should avoid using mutables as default values of a function.

Reply:
Good point, since we don't modify the `pmin_by_dict` or `pmin_by_type` parameters, it should be safe to use an object as the default value.

Reply:
Ah. Even if it will be altered only when it is empty?

Reply:
It seems safe here, but maybe we want to maintain the discipline to never do this so that we don't slip up when it matters.